Add scale kwarg to FlexAttention (and some changes that get FlexAttention numerics to be as accurate as FA2)

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130250
* #130227

After this PR, our numerical error is within 3% of FA2 for forward and gradients. Prior, for `dq` our numerical error was 30% higher. I also added a `PRESCALE_QK` kernel option that increases perf by about 3-4% but incurs about 20-30% more numerical error.

![image](https://github.com/pytorch/pytorch/assets/6355099/7b5ff44e-219b-4a05-8a1b-2a0182c01ab2)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Chillee committed:
Add scale kwarg to FlexAttention

[ghstack-poisoned]

vadimkantorov commented:
@Chillee would FlexAttention be made available as a backend for SDPA? E.g. will this enable FAv2 impl with custom `attn_bias`? (at least for forward pass)

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -121,51 +66,43 @@ def build_subgraph_buffer(
-            env[node] = (
-                args[lifted_input_index] if is_lifted_input else placeholder_inps[cnt]
-            )
+            env[node] = args[cnt]
```
I like this change, TLDR we made a mega list of [placeholders, then args] so that as we iterate we only get to the output node at the right arg count.

Can you update the docs, (remove args) and right some comment to the above effect.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -121,51 +66,43 @@ def build_subgraph_buffer(
-            )
-            return subgraph_buffer
+
+            def convert_output_node_to_buffer(output):
```
nit: I find inlined functions really hard to read especially if they are multiple levels of nested deep.. probs why I struggle with PT2 lol

Chillee replied:
I like inlined functions particularly in cases like this, because it keeps the definition of the function close to the actual usage.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -121,51 +66,43 @@ def build_subgraph_buffer(
+                if output is None:
+                    return None
+                output_node = output
+                output_buffer = env[output_node]
```
also why dont we need the node[0][0] is has the output been flattened? e.g. ahhh as I am writing this I realized why you tree_map

could we also bake this into some comment

drisspg commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -423,6 +424,7 @@ def flex_attention(
     value: torch.Tensor,
     score_mod: _score_mod_signature = _identity,
     block_mask: Optional[BlockMask] = None,
+    scale=None,
```
nit; add type hints please

drisspg commented on torch/nn/attention/_flex_attention.py:
```diff
@@ -454,6 +456,8 @@ def score_mod(
         value (Tensor): Value tensor; shape :math:`(B, H, S, Ev)`.
         score_mod (Callable): Function to modify attention scores. By default no score_mod is applied.
+        block_mask (BlockMask): BlockMask object that controls the blocksparsity pattern of the attention.
+        scale (float): Scaling factor applied prior to softmax. If none, the default value is set to :math`\frac{1}{\sqrt{E}}`
```
optional float

drisspg reviewed (approved):
left some comments, mostly nits

Chillee committed:
Update on "Add scale kwarg to FlexAttention"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Add scale kwarg to FlexAttention"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Add scale kwarg to FlexAttention (and some changes that get FlexAttention numerics to be as accurate as FA2)"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Add scale kwarg to FlexAttention (and some changes that get FlexAttention numerics to be as accurate as FA2)"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented:
@vadimkantorov We're still figuring out how to integrate FlexAttention with `F.sdpa`, but yes, FlexAttention allows you to provide arbitrary biases and attention masks.

Chillee commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge -i

izaitsevfb commented:
@pytorchbot revert -m "depends on #130227 which needs to be reverted" -c ghfirst

Chillee committed:
Update on "Add scale kwarg to FlexAttention (and some changes that get FlexAttention numerics to be as accurate as FA2)"


After this PR, our numerical error is within 3% of FA2 for forward and gradients. Prior, for `dq` our numerical error was 30% higher. I also added a `PRESCALE_QK` kernel option that increases perf by about 3-4% but incurs about 20-30% more numerical error.

![image](https://github.com/pytorch/pytorch/assets/6355099/7b5ff44e-219b-4a05-8a1b-2a0182c01ab2)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Add scale kwarg to FlexAttention (and some changes that get FlexAttention numerics to be as accurate as FA2)"


After this PR, our numerical error is within 3% of FA2 for forward and gradients. Prior, for `dq` our numerical error was 30% higher. I also added a `PRESCALE_QK` kernel option that increases perf by about 3-4% but incurs about 20-30% more numerical error.

![image](https://github.com/pytorch/pytorch/assets/6355099/7b5ff44e-219b-4a05-8a1b-2a0182c01ab2)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge

Chillee commented:
@pytorchbot merge
